core/src/apps/cardano/helpers/hash_builder_collection.py

```diff
@@ -0,0 +1,110 @@
+from apps.common import cbor
+
+if False:
+    from typing import Any, Generic, Generic as GenericKVP, TypeVar
+    from trezor.utils import HashContext
+
+    T = TypeVar("T")
+    K = TypeVar("K")
+    V = TypeVar("V")
+else:
+    T = 0  # type: ignore
+    K = 0  # type: ignore
+    V = 0  # type: ignore
+    Generic = [object]  # type: ignore
+    GenericKVP = {(0, 0): object}  # type: ignore
+
+
+class HashBuilderCollection:
+    def __init__(self, size: int) -> None:
+        self.size = size
+        self.remaining = size
+        self.hash_fn: HashContext | None = None
+        self.parent: "HashBuilderCollection" | None = None
+        self.has_unfinished_child = False
+
+    def start(self, hash_fn: HashContext) -> "HashBuilderCollection":
+        self.hash_fn = hash_fn
+        self.hash_fn.update(self._header_bytes())
+        return self
+
+    def _insert_child(self, child: "HashBuilderCollection") -> None:
+        child.parent = self
+        assert self.hash_fn is not None
+        child.hash_fn = self.hash_fn
+        self.has_unfinished_child = True
+        child.hash_fn.update(child._header_bytes())
+
+    def _do_enter_item(self) -> None:
+        assert self.hash_fn is not None
+        assert self.remaining > 0
+        if self.has_unfinished_child:
+            raise RuntimeError  # can't add item until child is finished
+
+        self.remaining -= 1
+
+    def _hash_item(self, item: Any) -> None:
+        assert self.hash_fn is not None
+        for chunk in cbor.encode_streamed(item):
+            self.hash_fn.update(chunk)
+
+    def _header_bytes(self) -> bytes:
+        raise NotImplementedError
+
+    def finish(self) -> None:
+        if self.remaining != 0:
+            raise RuntimeError  # not all items were added
+        if self.parent is not None:
+            self.parent.has_unfinished_child = False
+        self.hash_fn = None
+        self.parent = None
+
+    def finish_and_get_hash(self) -> bytes:
+        assert self.hash_fn is not None
+        hash = self.hash_fn.digest()
+
+        self.finish()
+
+        return hash
+
+    def __enter__(self) -> "HashBuilderCollection":
+        assert self.hash_fn is not None
+        return self
+
+    def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
+        if exc_val:
```

Comment:
exception will fall through if `__exit__` returns a falsey value, so this is a very weird thing to do explicitly

if you want to make sure that `finish()` is not called in case of exception, do instead:
```
if exc_type is None:
	self.finish()
```

Reply:
Can `exc_val` be a falsey value? I think I need to do:
```
if exc_val is not None:
    raise exc_val
    
self.finish()
```
I need this, because otherwise the errors thrown inside the `with` statement are never propagated which causes tests to fail because of invalid errors - the errors are simply `RuntimeError`s then and also the stack trace isn't helpful for development then.

Reply:
```python3
class Ctx:
    def __enter__(self):
        return self

    def __exit__(self, a, b, c):
        if a is None:
            raise RuntimeError("this is from finish()")


with Ctx():
    raise RuntimeError("this is from outside the class")
```

```
Traceback (most recent call last):
  File "test.py", line 11, in <module>
    raise RuntimeError("this is from outside the class")
RuntimeError: this is from outside the class
```

Reply:
🤦 Aaah, I see now. I was confused by what's happening. It was indeed failing because of an error being thrown inside of `finish`. My bad, sorry. https://github.com/trezor/trezor-firmware/pull/1683/commits/90a0ad9192bd9ab7cc3e3523526ea6bf376c2ba9